### PR TITLE
Lair compiler steps

### PR DESCRIPTION
**Pull request by gabriel-barrett**
This PR substitutes `check_and_link` by a 3-staged approach. First a check, then an expansion (translates complex operations into simple ones), then a compilation into bytecode. This approach, although slightly more inefficient, is much more principled and robust

PS: the reason the traces of if statements changed is that the true/false selectors flipped. In the original, it was identifying the true branch as the first branch, now it is the false which is the first. The reason is that the ifs are translated into matches and the true branch is the default case, which appears last

### CODE REVIEW

File: src/lair/toplevel.rs [hunk 1]
```diff
@@ -189,18 +267,30 @@ impl<F: Field + Ord> FuncE<F> {
...
+
+    fn compile<C1: Chipset<F>, C2: Chipset<F>>(&self, ctx: &mut LinkCtx<'_, C1, C2>) -> Block<F> {
+        let mut ops = vec![];
+        self.ops.iter().for_each(|op| op.compile(&mut ops, ctx));
         let ops = ops.into();
         let saved_return_idents = std::mem::take(&mut ctx.return_idents);
-        let ctrl = self.ctrl.check_and_link(ctx);
+        let ctrl = self.ctrl.compile(ctx);
         let block_return_idents = std::mem::take(&mut ctx.return_idents);
+        // sanity check
```
Comment (wwared) [inline src/lair/toplevel.rs:293]: nit: Can this sanity check (the assert below this comment) be moved to the `check` function on `BlockE` instead?

**Review by wwared — DISMISSED** [review]
Looks like a fine refactor to me. Small nit below

**Review by arthurpaulino — APPROVED** [review]
As the code approaches stability and correctness, we should be striving for more docstrings, especially on more involved functions.

`check`, `expand` and `compile` would be perfect candidates for this. Also, due to the relevance of `toplevel.rs`, a module docstring would be really good.

**Review by arthurpaulino — APPROVED** [review]
Thanks a bunch!

**Pull request merged**